### ISSUE TITLE
Storage corruption when addition on storage element overflows

### ISSUE THREAD

**Issue by catageek** [ethereum/solidity#1306]
The following code returns 1 instead of 0 because the addition overflows. The assembly code shows that the AND mask is applied before the addition and not after, resulting in the corruption of b.

```
contract Test {
    uint32 a = 0xFFFFFFFF;
    uint32 b;

    function run() returns (uint32) {
        var x = 1;
        a = uint32(a + x);
        return b;
    }
}
```

I suppose that writing in a should not corrupt b in any case.

Solidity version: 0.4.3+commit.2353da71.Emscripten.clang

**Comment by ethernomad** [comment]
Confirmed. I can reproduce this bug with and without optimization all the way back to 0.1.6.

**Comment by chriseth** [comment]
Simpler example:

```
contract Test {
    uint32 a;
    uint32 b;

    function run() returns (uint32) {
        a--;
        return b;
    }
}
```

**Comment by ghost** [comment]
Has this issue been fixed if we do pragma solidity ^0.4.4?

**Comment by axic** [comment]
@Physes yes, it is fixed in the 0.4.4 version and the pragma will ensure it cannot be compiled with earlier versions.

**Comment by VIP21** [comment]
#4674
